src/visual.ts

```diff
@@ -329,65 +355,188 @@ module powerbi.extensibility.visual {
             };
         }
 
+        // in this method we breaking simple cycles for typical displaying with twice rendering onr node in cycle
+        private processCycles(cycles: SankeyDiagramCycleDictionary, nodes: SankeyDiagramNode[], links: SankeyDiagramLink[]): SankeyDiagramLink[] {
+            let createdNodes: SankeyDiagramNode[] = [];
+            for (let nodeName in cycles) {
+                let firstCyclesNode: SankeyDiagramNode = cycles[nodeName].filter((node: SankeyDiagramNode): boolean => {
+                    if (node.label.name === nodeName) {
+                        return true;
+                    }
+                    return false;
+                })[0];
+
+                let nodeCopy: SankeyDiagramNode = _.cloneDeep(firstCyclesNode);
+                nodeCopy.label.name += SankeyDiagram.DublicatedNamePostfix;
+
+                // copy only! output links to new node;
+                nodeCopy.links = firstCyclesNode.links.filter((link: SankeyDiagramLink) => {
+                    if (link.source === firstCyclesNode || link.source === link.destination) {
+                        return true;
+                    }
+                    return false;
+                });
+                nodeCopy.links.forEach((link: SankeyDiagramLink) => {
+                    link.source = nodeCopy;
+                });
+
+                // remove output links from original node
+                firstCyclesNode.links = firstCyclesNode.links.filter((link: SankeyDiagramLink) => {
+                    if (link.destination === firstCyclesNode || link.destination === link.source) {
+                        return true;
+                    }
+
+                    return false;
+                });
+
+                SankeyDiagram.updateValueOfNode(firstCyclesNode);
+                SankeyDiagram.updateValueOfNode(nodeCopy);
+                SankeyDiagram.fixLinksCount(firstCyclesNode);
+                SankeyDiagram.fixLinksCount(nodeCopy);
+                nodes.push(nodeCopy);
+                createdNodes.push(nodeCopy);
+            }
+
+            return links;
+        }
+
+        // remove dublicated links
+        private static fixLinksCount(node: SankeyDiagramNode) {
+            node.links = _.uniq(node.links);
+        }
+
+        // Depth-First Search 
+        private checkCycles(nodes: SankeyDiagramNode[]): SankeyDiagramCycleDictionary {
+            let nodesStatuses: any = [];
+
+            // init nodes statuses array
+            // all nodes are not visited state
+            nodes.forEach((node: SankeyDiagramNode) => {
+                if (node.links.length > 0) {
+                    nodesStatuses[node.label.name] = {
+                        node: node,
+                        status: SankeyDiagramNodeStatus.NotVisited
+                    };
+                }
+            });
+
+            let simpleCycles: SankeyDiagramCycleDictionary = {};
+
+            let dfs = (node: SankeyDiagramNode) : void => {
```

Comment:
I'd prefer to declare a method in order to have an ability to test it. Anonymous function aren't good for UT